### DISCUSSION TITLE
Smart Border Crop for Frames without loosing external pieces

### DISCUSSION THREAD

**Discussion by LostOnTheLine** [ImageMagick/ImageMagick#5867] [Help]
So I am trying to make an auto-processing script that uses ImageMagic to process Manga (Comic) like pages in the back of a number of books.
The Images typically have a white border that goes around the whole page usually leaving between ⅕ & ¼ of the page for wasted blank space. There are, however quite often things that stick out of the frame like text that might sit above or below a frame or a banner that goes across a corner or even a head or arm of a character that sticks out a little. The main image is almost always 8 frames in a 2x4 arrangement. but sometimes 2 or 4 of those frames are together for a larger frame but with the same edges.

I've tried this before with other software but the result were always bad enough to not be worth it. But ImageMagick seems to have a lot more smarts & capability & customizability, so I'm hoping to be able to make it work.

Some things that are important

- I want there to still be a border around the images, just a much smaller one
- I want the border to be centered with the frame area but not cutting off any of the things that stick out
- The images will all be downscaled after, so having them all exactly even is not important & but I'd prefer them to at least be close
- The plan is to use this is a script so it can be applied to many images sequentially so a visual inspection &/or adjustment is not really possible

Some examples of images I have manually cropped to decent versions (Blurred as to not be a copyright issue)
Image 1 I think has a good border & a balanced feel while you can see the part sticking up in the Upper Left corner, it still has a border that's big enough to leave a border around the parts that stick out, but is centered accross the frame lines not the furthest non-white spot
Image 2 is a bit lop-sided, but it is good enough that if it was what could be done in an automation I'd be happy with it
Image 3's border is, a little too small to be idea, but still what I'd call acceptable
Image 4 

<img src="https://user-images.githubusercontent.com/60684934/206597499-318f295e-46ce-44ac-9154-3e86ce6e72de.jpg" width="350" /> | <img src="https://user-images.githubusercontent.com/60684934/206597665-e1e55ff5-5079-40cc-9f55-02bc81b1a3e3.jpg" width="350" />

⚫ ⚫ ⚫ ⚫ ⚫ **Image 1: _Original_** ⚫ ⚫ ⚫ ⚫ ⚫ ⚫ ⚫ ⚫ **Image 1: _A Good Cropped Version_** ⚫ ⚫ ⚫

<img src="https://user-images.githubusercontent.com/60684934/206597921-5621431c-6a82-403e-b667-fab63bdaed82.jpg" width="350" /> | <img src="https://user-images.githubusercontent.com/60684934/206597959-cf95b45f-d3c4-410a-a9f1-bb81b76205ae.jpg" width="350" />
⚫ ⚫ ⚫ ⚫ ⚫ **Image 2: _Original_** ⚫ ⚫ ⚫ ⚫ ⚫ ⚫ ⚫ ⚫ **Image 2: _A Decent Cropped Version_** ⚫ ⚫ ⚫

<img src="https://user-images.githubusercontent.com/60684934/206598038-580754c9-6c77-48ef-976c-f10f27d2c0d7.jpg" width="350" /> | <img src="https://user-images.githubusercontent.com/60684934/206598081-c0e56ccc-a039-4b42-b83f-4fa816812ca7.jpg" width="350" />
⚫ ⚫ ⚫ ⚫ ⚫ **Image 3: _Original_** ⚫ ⚫ ⚫ ⚫ ⚫ ⚫ ⚫ ⚫ **Image 3: _Okay Cropped Version_** ⚫ ⚫ ⚫ ⚫

<img src="https://user-images.githubusercontent.com/60684934/206598124-5da3948f-3640-4d8c-a92c-e4a2a13a61a9.jpg" width="350" /> | <img src="https://user-images.githubusercontent.com/60684934/206598150-1efb94f6-d185-4c51-b6ad-e7476d42e464.jpg" width="350" />
⚫ ⚫ ⚫ ⚫ ⚫ **Image 4: _Original_** ⚫ ⚫ ⚫ ⚫ ⚫ ⚫ ⚫ ⚫ **Image 4: _Acceptable Cropped Version_** ⚫ ⚫

------------------------
--------------------------
------------------

So the ideal I'm aiming for is along these lines (Pun intended) 

- With the Red line being the edged of the clearly defined large vertical & horizontal edges of the frames
- & the Blue line being the furthest outside point that is not white.

**### Is there a way to have ImageMagick find those values & use them for a relative points for a `-trim` argument?** 
I figure it'd be along the lines of figuring out a happy padding value from blue then taking the distance from Blue+Padding to Red & keeping that value around the whole thing or at least on the sides & maybe doing a different one on Top/Bottom, while having some sort of maximum distance from blue also set.

<img src="https://user-images.githubusercontent.com/60684934/206601195-4777e389-61e6-4c8e-a264-265fa73d27df.jpg" data-canonical-src="https://user-images.githubusercontent.com/60684934/206601195-4777e389-61e6-4c8e-a264-265fa73d27df.jpg" width="400" />

**Comment by snibgo** [comment]
What version of IM, on what platform?

This command ...
```
magick in.ext -format %@ info:
```
... outputs the parameters to "`-crop`" that a trim would give, in this case trimming the white border so some non-white pixels will touch each edge. This is your blue line.

The four numbers are WxH+X+Y: width, height, x-offset and y-offset of the crop. A shell script can read those four numbers and do whatever you want.

You can use "-fuzz" if necessary.

This command ...
```
magick in.ext -define trim:percent-background=10% -format %@ info:
```
... gives the crop that would leave just 10% of white pixels at each edge. This is your red line. See https://www.imagemagick.org/discourse-server/viewtopic.php?f=4&t=35579&sid=72010b9693694a55148254e901578425

**Reply by LostOnTheLine** [reply]
> What version of IM, on what platform?

I'm running `Version: ImageMagick 7.1.0-53 Q16-HDRI x64 b0cc8e4:20221204` on `Windows 10` being called through a python script in a plugin for Sigil

> This command ...
> 
> ```
> magick in.ext -format %@ info:
> ```
> 
> ... outputs the parameters to "`-crop`" that a trim would give, in this case trimming the white border so some non-white pixels will touch each edge. This is your blue line.

But that's not the blue line, it's the Blue Line for the Upper Left & the Red for the Bottom Right. That's exactly the problem that I was trying to fix.
I'll post some better images to make it easier to understand hopefully

Here are some cleaned instead of blurred & a couple re-created with things that stick out so the lines are easier to see
<img src="https://user-images.githubusercontent.com/60684934/206909382-f7e92539-bda0-4147-845d-85f73db4b468.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/206909383-2496bf3f-694a-4094-a326-ce5266ee3e1d.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/206909384-08969727-bcaa-4aae-be94-17f8ea72919c.jpg" width="200" />

This is what happens when they get trimmed with `magick.exe mogrify -quality 90 -trim -fuzz 1% -filter Lanczos -rotate -90> -strip -resize 800x1100`. notice the unbalanced problem I was trying to illustrate with the colored lines
<img src="https://user-images.githubusercontent.com/60684934/206982049-d2af86ff-f964-4280-8340-d77cac8c7adb.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/206982052-aa38717c-c69b-4eea-95a9-b04e45bb7762.jpg" width="200" />

& this is what happens when I try to add the `-define trim:percent-background=10%` to it with `magick.exe mogrify -quality 90 -define trim:percent-background=10% -trim -fuzz 1% -filter Lanczos -rotate -90> -strip -resize 800x1100`. I'm including one of the blurred because it's even worse than the samples
<img src="https://user-images.githubusercontent.com/60684934/206982223-1ca8ba48-94e8-40e2-ac90-515f6b324f4e.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/206982225-a7854d18-0abf-44c5-a6ac-a7493fce5c33.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/206984981-ceb1ebc4-845a-40f3-a79e-e14575c9a0bd.jpg" width="200" />


> The four numbers are WxH+X+Y: width, height, x-offset and y-offset of the crop. A shell script can read those four numbers and do whatever you want.
> 
> You can use "-fuzz" if necessary.
> 
> This command ...
> 
> ```
> magick in.ext -define trim:percent-background=10% -format %@ info:
> ```
> 
> ... gives the crop that would leave just 10% of white pixels at each edge. This is your red line. See https://www.imagemagick.org/discourse-server/viewtopic.php?f=4&t=35579&sid=72010b9693694a55148254e901578425

I get this error `magick.exe: unable to open image 'in.ext': No such file or directory @ error/blob.c/OpenBlob/3570.` when attempting to run that in the CLI so I don't think I have that. Without the `in.ext` & the `-format %@ info` It runs, but it seems to go the other way, it outputs the images with MORE cropped off, not less. I'm fairly confident that I can find a way to increase the border, so I wasn't worrying about that until I found a way to adjust the detected borders as that would make the solutions that I've found, cropping the image then adding a border back, easier. If I can do it without cropping then re-adding all the better, but the major problem is still the centering. Some are more extreme than this one with borders similar to this in the more extreme cases

<img src="https://user-images.githubusercontent.com/60684934/206982758-321161aa-8b01-4c23-a432-62a6b23eb0a1.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/206984525-06c129f7-55c8-4e5d-ae17-56329a16dcc8.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/206984544-f76e981e-1a07-4a78-ad6f-881df3864bb3.jpg" width="200" />

I'm happy to know that there's an option for leaving a border, that's much better than adding it back IMO, but I'll have to figure out how to get it to work properly, but the big issue is still how bad they look that you can see even without the border

**Comment by LostOnTheLine** [comment]
I've worked out a way to get it to give me a border that actually works.
Unfortunately it requires that I know the image dimensions because I need to give a value for `trim:minSize=` having to do with the image before it is resized. On the book I'm using for testing I manually measured a good shape & used those dimensions. The original Image was `1400x1948` & I found `1200x1700` created a good amount of border. so I used 

```
magick.exe mogrify -quality 90 -fuzz 1% -define trim:minSize=1200x1700 -trim -filter Lanczos -rotate -90> -strip -resize 800x1100
```
<img src="https://user-images.githubusercontent.com/60684934/207017979-fa3264ae-8656-4243-b287-4f2b6d0ed222.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207017981-b0bfdef9-bcf1-4608-877d-05f3a8ebb7f1.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207018959-09a4dfd5-d34f-4677-93d0-6339706a3719.jpg" width="200" />


The results had everything cut close at the top leaving a bunch of white on the lower right, so I had to add `-gravity Center` to make it center the crop & it looks decent, but at least has a good proportion, if not centered. on the test image with the extreme popouts it still pressed it to the limit, but that can't be helped really. But having to specify the dimensions makes it not really able to be easily automated unfortunately. I tried `90%` for the `-define trim:minSize=1200x1700` but it just acts as though I didn't define the `minSize=` at all. I would think it was the `%` needing to be escaped, except the `%` worked fine in the `percent-background=10%`. Not sure what I need to do to get that to work. With the gravity  I feel there should be a way to have the -trim crop the same amount on each side so if it can only crop 20 pixels off the left it will stop at 20 pixels on the right even though it could crop 60. I feel this is a simpler method than having it detect the lines with a seam carving-type algorithm since if it trims the same amount off each side that should keep it as centered as it initially was. now just need to figure out how...

this is my current working set
```
magick.exe mogrify -quality 90 -fuzz 1% -gravity Center -define trim:minSize=1200x1700 -trim -filter Lanczos -rotate -90> -strip -resize 800x1100
```

<img src="https://user-images.githubusercontent.com/60684934/207018145-984cf937-ef9b-49de-aa4c-2fc1a04f1295.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207018147-8c3760c0-8539-45b7-bfd6-db525a34e49c.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207018716-7910f9db-7638-4bd2-941a-69a22f2de005.jpg" width="200" />

**Comment by snibgo** [comment]
> I get this error magick.exe: unable to open image 'in.ext': No such file or directory @ error/blob.c/OpenBlob/3570.

IM tells you it can't find a file called "in.ext". Use the actual name of your input file in place of "in.ext".

> Here are some cleaned instead of blurred & a couple re-created with things that stick out so the lines are easier to see

Thanks for those. But you haven't shown what results you want from those inputs, so I have to guess.

I assume you want the output image to contain all the dark pixels from the input.

I assume you want to remove the same amount on the left and right, but not necessarily the same amount at top and bottom.

I assume you want to leave an extra 10 pixels of white border on all edges.

I assume you don't want to remove "TEXT TEXT" from the top of the first image.

The inputs are JPG, and the whites are not exactly white, so we need a fuzz factor.

I know nothing about Python or Sigil. Below is a Windows BAT script, named `forTrim.bat`. I will leave translation to Python to you.
```
set INFILE=%1
set OUTFILE=%2

for /F "usebackq" %%L in (`%IMG7%magick ^
  %INFILE% ^
  -format "WW=%%w\nHH=%%h\n" +write info: ^
  -fuzz 10%% -format "JustWhite=%%@\n" +write info: ^
  -define "trim:percent-background=80%%" -format "Inner=%%@\n" +write info: ^
  NULL:`) do set %%L

rem Below, we don't use "Inner".

%IMG7%magick ^
  %INFILE% ^
  ( +clone -crop %JustWhite% +repage -write jw1.jpg +delete ) ^
  ( +clone -crop %Inner% +repage -write inner1.jpg +delete ) ^
  NULL:

rem Calculate CL, CR, CT, CB as amounts we might crop from left, right, top and bottom.
for /F "tokens=1-4 delims=x+" %%A in ("%JustWhite%") do (
  echo %%A %%B %%C %%D

  set /A CL=%%C
  set /A CR=%WW%-%%A-%%C
  set /A CT=%%D
  set /A CB=%HH%-%%B-%%D
)

echo CL=%CL% CR=%CR% CT=%CT% CB=%CB%


rem "Extra" is extra pixels for an output border.
set Extra=10


rem Calculate CLR as minimum of CL and CR; CTB as minimum of CT and CB
set CLR=%CL%
if %CLR% GTR %CR% set CLR=%CR%

echo CLR=%CLR%

set /A CLR-=%Extra%
set /A CT-=%Extra%
set /A CB-=%Extra%

echo CLR=%CLR%  CT=%CT% CB=%CB%


rem Calculate new width and height.
set /A NW=%WW%-2*%CLR%
set /A NH=%HH%-%CT%-%CB%

echo NW=%NW% NH=%NH%

rem Do the final crop.

%IMG7%magick ^
  %INFILE% ^
  -crop %NW%x%NH%+%CLR%+%CT% ^
  +repage ^
  %OUTFILE%
```

Call the script like this, from a Windows command window:
```
call forTrim forTrim1.jpg out1.jpg
call forTrim forTrim2.jpg out2.jpg
call forTrim forTrim3.jpg out3.jpg
```
The three outputs from your inputs are:

![out1](https://user-images.githubusercontent.com/33812368/207068367-0172d3bf-151a-44a4-b5d0-976ccb211692.jpg)
![out2](https://user-images.githubusercontent.com/33812368/207068390-0eaf3ebd-cbc8-4784-8564-83d7a89415bd.jpg)
![out3](https://user-images.githubusercontent.com/33812368/207068413-8d74a32d-f568-4bfe-b414-82b781516ba1.jpg)

**Reply by LostOnTheLine** [reply]
> > I get this error magick.exe: unable to open image 'in.ext': No such file or directory @ error/blob.c/OpenBlob/3570.
> 
> IM tells you it can't find a file called "in.ext". Use the actual name of your input file in place of "in.ext".

Ah, that makes sense. The calls I see don't seem to have the image at the front so I wasn't expecting that to be the image.

> > Here are some cleaned instead of blurred & a couple re-created with things that stick out so the lines are easier to see
> 
> Thanks for those. But you haven't shown what results you want from those inputs, so I have to guess.

Oh. I'll add some, I did with the blurred, but I'll add some with the samples I made.
> I assume you want the output image to contain all the dark pixels from the input.
I would like it to be flexible enough to handle it if it was black background as well, but yes, I want all pixels that are not the background color to be preserved.
> I assume you want to remove the same amount on the left and right, but not necessarily the same amount at top and bottom.
> I assume you want to leave an extra 10 pixels of white border on all edges.

Since it's doing this before resizing I don't really want to rely on pixel count because then if the initial image is 3000 & is scaled down to 1000 those 10 pixels will end up as 3 & if it starts as 1400 & scales down to 1000 it will end up with a 7, a much bigger border being more than twice as big. Currently my working set seems to look nice with up to 10-15% of the total image size removed if it's background, but in a perfect world I'd probably have them as 8-10% on the sides & maybe 12-15% on the top, but once I have a way to do it I can work with testing a number of different books to find a good one that looks best.
> I assume you don't want to remove "TEXT TEXT" from the top of the first image.

Definitely Not
> The inputs are JPG, and the whites are not exactly white, so we need a fuzz factor.

So from my testing all of the Manga/Comics worked good with a 1% Fuzz, the Maps I'm currently using 18% because the ones I've tested all worked with that, but I expect that may need to change or possibly be set as a User-Set variable.
> I know nothing about Python or Sigil. Below is a Windows BAT script, named `forTrim.bat`. I will leave translation to Python to you.

That's perfectly fine. I'm not actually that well versed in Python myself, I used it a little forever ago, but never really learned it. Now pretty much everything I've learned is antiquated & there are much better ways.
I actually have been converting all the arguments to a simple CLI format because Python handles them much differently & I figured it'd be easier to present everything in a way that people here will be most familiar with
In Python the formatting actually looks like this. I believe because it uses the `mogrify` it just sets the output image as the input image as well.
```
['C:/Program Files/ImageMagick-7.1.0-Q16-HDRI/magick.exe', 'mogrify', '-quality', '90', '-fuzz', '1%', '-gravity', 'Center', '-define', 'trim:minSize=1260x1753', '-trim', '-filter', 'Lanczos', '-rotate', '-90>', '-strip', '-resize', '800x1100', '-path', 'R:\\TEMP\\USERtemp', 'R:\\TEMP\\USERtemp\\Sigil-lHMPPp\\OEBPS\\Images\\Mangas2.jpg']
```

-----
-----

Here are the non-blurred images with The Original, A marked version, & what I would call a god cut.
<img src="https://user-images.githubusercontent.com/60684934/207093119-4e7c5690-7698-483b-9f14-48da075bc5f4.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207093126-2c8b9de0-7813-4170-84a9-a7f5f6b13270.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207093123-f2d5f152-450e-4711-9853-d08cc40409b2.jpg" width="200" />
---
<img src="https://user-images.githubusercontent.com/60684934/207093161-b596924e-7067-4c6c-bf68-a55aa7351a8d.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207093170-76aa5997-e48d-4cf6-8f36-f6de9d885950.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207093165-6f643969-a627-48bf-9e67-f4dca4f07e66.jpg" width="200" />
---
<img src="https://user-images.githubusercontent.com/60684934/207093202-6e23fd8b-8e65-4ac8-871f-8f6547a74194.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207093212-5285d82c-fd8c-42ca-a5e5-1c2011f5ba05.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207093206-badf8d1a-bc0d-49a2-83f0-d6127375d413.jpg" width="200" />
---

**Comment by LostOnTheLine** [comment]
So, for now, I'm having python detect the image dimensions from the file, then do a 90% integer from those measurements & use that, with the center, for the Manga/Comics border trim. It's not centered, but it's at least a workable solution.

I'll post the Sigil Extension on [MobileReads](https://www.mobileread.com/forums/showthread.php?p=4186008) Hopefully someone more skilled than I can clean up my hacked-together code, but it does work.

It currently offers the following options to be applied to images in an ePUB

- "maxWidth": The Max Width all images larger than will be downscaled to `-resize {maxWidth}x{maxHeight}`
- "maxHeight": The Max Height all images larger than will be downscaled to
- "qualityNum": The Quality all images being downscaled will be saved as `-quality {qualityNum}`
- - JPEG= Quality Level 1-100
- - PNG= Compression Level <10=0,10-19=1,20-29=2,90-99=9
- "tallImages": If set to True will rotate all images that are wide 90° Counterclockwise `-rotate -90>`
- "trimMaps": If set to True will trim borders on all images with the word "map" in the name with `-fuzz 18% -trim`
- "trimComics": If set to True will trim borders on all images with the word "comic" or "manga" in the name with `-fuzz 1% -gravity Center -define trim:minSize={int(w*0.9)}x{int(h*0.9)} -trim`
- "customArg": A place to put any custom arguments to be added to the commands for all images. 
- - All `,` & ` ` are treated as argument separators so `-fuzz_____1%,_-gravity,__Center` will end up as `-fuzz_1%_-gravity_Center`
- "stripComments": If set to True will remove any comments or text chunks from images

The Full path for an image if everything was called, (with Map Arguments replacing Comic ones for Maps)
```
magick_exe_path mogrify -quality {qualityNum} -fuzz 1% -gravity Center -define trim:minSize={int(w*0.9)}x{int(h*0.9)} -trim {customArg} -rotate '-90>' -strip -resize {maxWidth}x{maxHeight} -path temp_dir original_img_path
```

@snibgo The `trim:percent-background=` didn't work but it led me to the `trim:minSize=` so thank you for that.
I'm still trying to get the frames to be centered, I'm thinking my best bet is to try to get it to detect the `Blue Line` distance, remove some % of that, with a max somewhere, & have it do equal trims on all sides relative to that measurement.

Also, if there's any other arguments someone thinks should be options for an ePUB please let me know. It started with only the option to set the maxWidth & nothing else, so I've been adding things I think would be useful. I expect more feedback from the MobileReads community on that, but those in the ImageMagick Community may have ideas because they know the capabilities of ImageMagick & I appreciate any help I can get.
